galley/pkg/mcp/snapshot/inmemory.go

```diff
@@ -0,0 +1,49 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package snapshot
+
+import mcp "istio.io/api/config/mcp/v1alpha1"
+
+// InMemory Snapshot implementation
+type InMemory struct {
```

Comment:
This should be protected with a `sync.Mutex`.

Reply:
Given that this is semantically immutable, do we need sync.Mutex?

Reply:
The current code suggests a mutable in-memory snapshot with no locking. We either ought to:

- add locking and allow it to be safely mutable, relying on convention for immutability after snapshot is set. An explicit `Freeze()` method could prevent mutation.

or 

- remove `Set()` and enforce immutability at object creation time, i.e. pass map of resources and verions to `New()`.

Reply:
Added Freeze() and made Set() panic when frozen.